### PR TITLE
Sans-IO WAMP Dealer and Broker

### PR DESCRIPTION
**Pull request by Mahad-10**

### CODE REVIEW

File: lib/src/dealer.dart [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      }
+
+      var registrations = registrationsBySession[sessionID];
+      if (registrations == null || !registrations.containsKey(message.registrationID)) {
+        throw Exception("registration for session $sessionID doesn't exist");
+      }
+      var procedure = registrations[message.registrationID];
+
+      registrationsBySession[sessionID]!.remove(message.registrationID);
+      registrationsByProcedure[procedure]!.remove(message.registrationID);
```
Comment (muzzammilshahid) [inline lib/src/dealer.dart:89]:
```suggestion
      registrationsBySession[sessionID]?.remove(message.registrationID);
      registrationsByProcedure[procedure]?.remove(message.registrationID);
```

File: lib/src/types.dart [hunk 2]
```diff
@@ -6,3 +8,9 @@ class SessionDetails {
   final String authid;
   final String authrole;
 }
+
+class MessageWithRecipient {
+  MessageWithRecipient(this.message, this.recipient);
```
Comment (muzzammilshahid) [inline lib/src/types.dart:13]: Please add empty line after this

**Review by muzzammilshahid — CHANGES_REQUESTED** [review]
Suggested some changes in dealer regarding null safety. Similar changes need to be done in broker as well

File: lib/src/dealer.dart [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      Result result = Result(message.requestID, args: message.args, kwargs: message.kwargs);
+      return MessageWithRecipient(result, sessionID);
+    } else if (message is Register) {
+      if (!registrationsBySession.containsKey(sessionID)) {
+        throw Exception("cannot register, session $sessionID doesn't exist");
+      }
+
+      int registrationID = idGen.next();
+      registrationsByProcedure[message.uri] = {registrationID: sessionID};
+      registrationsBySession[sessionID] = {registrationID: message.uri};
```
Comment (muzzammilshahid) [inline lib/src/dealer.dart:73]: Same here

**Review by muzzammilshahid — APPROVED** [review]

**Review by om26er — APPROVED** [review]

**Pull request merged**